### PR TITLE
gen4: Support explicit column aliases on derived tables

### PR DESCRIPTION
**Pull request by systay**
## Description
This PR adds support for derived tables that have column aliases explicitly stated listed as part of the derived table alias, such as this query:

```sql
select * 
from (
	select id, count(*) 
	from user 
	group by id
) as dt(user_id, user_count);

+---------+------------+
| user_id | user_count |
+---------+------------+
...
+---------+------------+
```

The PR also finished the moving of the planner to the operator data structures, removing a lot of now unused code.

## Related Issue(s)
Tracking issue https://github.com/vitessio/vitess/issues/11626

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/union_cases.json [hunk 1]
```diff
@@ -749,22 +749,29 @@
...
       "Instructions": {
-        "OperatorType": "Distinct",
-        "Collations": [
-          "(0:1)"
+        "OperatorType": "SimpleProjection",
+        "Columns": [
+          0
         ],
-        "ResultColumns": 1,
         "Inputs": [
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/union_cases.json:756]: good fix.

File: go/vt/vtgate/planbuilder/testdata/union_cases.json [hunk 2]
```diff
@@ -1083,35 +1098,41 @@
...
-            "OperatorType": "Distinct",
-            "Collations": [
-              "(0:1)"
-            ],
+            "OperatorType": "Projection",
+            "Expressions": null,
             "Inputs": [
               {
-                "OperatorType": "Concatenate",
+                "OperatorType": "Distinct",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/union_cases.json:1105]: Looks like this Projection is not needed.
Compact can mostly merge the two projection operators here.

**Review by harshit-gangal — APPROVED** [review]
overall this looks good. Some comments were added.
Approving it as it can be merged as-is as well.

File: go/vt/vtgate/planbuilder/projection.go [hunk 3]
```diff
@@ -41,28 +40,8 @@ var _ logicalPlan = (*projection)(nil)
...
-		if err != nil {
-			return err
-		}
-		columns = append(columns, convert)
-	}
-	p.primitive = &engine.Projection{
-		Cols:  p.columnNames,
-		Exprs: columns,
+	if p.primitive == nil {
+		return vterrors.VT13001("should already be done")
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/projection.go:44]: Why not a panic here?

**Review by GuptaManan100 — APPROVED** [review]
LGTM overall

**Pull request merged**